open-craft/edx-platform
# [SE-3401] Updates BuildTools UglifyJS configuration to be consistent on different machines

When generating the minified files for the TinyMCE static files using the `vendor_extra/tinymce/JakePackage.zip`, there's often a problem that was happening where the files are different after running the following command:
```
jake minify bundle[themes:modern,plugins:advlist,anchor,autolink,charmap,code,codemirror,contextmenu,image,insertdatetime,link,lists,media,paste,print,save,searchreplace,table,textcolor,visualblocks]
```

This becomes an issue when installing a new TinyMCE plugin on different machines and re-running `jake minify bundle`, because the files are not consistent on different machines. This results in a different `commons.js` file hash on the different machines, which makes it not possible to use multiple instances with a load balancer.

There are other reasons why the `commons.js` file hash is different, which are mentioned in PR edx#24972
Therefore, **this is a follow-up PR to #24972**.

This PR simply adds a specific configuration for `uglify-js`, which is:
```
mangle : {
	toplevel: false,
	sort: true
},
no_functions : false,
ascii_only: true,
compress: {
	cascade: false,
	comparisons: false,
	if_return: true
}
```

One of the most important is, `sort` which makes the mangled variable names consistent on different runs. Also, `comparisons` and `if_return` are specified because they were sometimes applied, while others times, they weren't. 

**JIRA tickets**: SE-3401, SE-3101, SE-3381

**Upstream PR**: edx#25324

**Testing instructions**:

1. Change your directory to `common/static/js/vendor/tinymce/`
2. Once you are there, run `unzip ../../../../../vendor_extra/tinymce/JakePackage.zip`
3. Clear the existing using `rm -rf node_modules`
4. Install the npm dependencies using `npm install` 
5. Run the jake command, mentioned below.
6. Commit the changes since I didn't update the existing minified files.
7. Run the jake command, mentioned below, again.
8. Make sure no changes have been made to the files.

```
jake minify bundle[themes:modern,plugins:advlist,anchor,autolink,charmap,code,codemirror,contextmenu,image,insertdatetime,link,lists,media,paste,print,save,searchreplace,table,textcolor,visualblocks]
```

**Author notes and concerns**:

1. Do you think it is better to update the minified files and commit those changes? Please let me know.

**Reviewers**
